src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java

```diff
@@ -0,0 +1,135 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.aurora.scheduler.updater.strategy;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Ordering;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A strategy that limits the number of instances selected by the subclass using variable steps.
+ *
+ * @param <T> Instance type.
+ */
+public class VariableBatchStrategy<T extends Comparable<T>> implements UpdateStrategy<T> {
+  private final Ordering<T> ordering;
+  protected final ImmutableList<Integer> maxActiveGroups;
+  private final boolean rollingForward;
+  private Optional<Integer> instanceModificationCount;
+
+  private static final Logger LOG = LoggerFactory.getLogger(VariableBatchStrategy.class);
+
+  /**
+   * Creates a variable active-limited strategy that applies an upper bound to all results.
+   *
+   * @param maxActiveGroups  List of Maximum group sizes. Each group size represents a step.
+   * {@link #getNextGroup(Set, Set)}.
+   */
+  public VariableBatchStrategy(Ordering<T> ordering,
+      List<Integer> maxActiveGroups,
+      boolean rollingForward) {
+
+    this.ordering = Objects.requireNonNull(ordering);
+    this.rollingForward = rollingForward;
+
+    maxActiveGroups.forEach(x -> Preconditions.checkArgument(x > 0));
+
+    this.maxActiveGroups = ImmutableList.copyOf(maxActiveGroups);
+    this.instanceModificationCount = Optional.empty();
+  }
+
+  private int determineStep(int idle) {
+
+    // Calculate which step we are in by finding out how many instances we have left to update.
+    int scheduled = instanceModificationCount.get() - idle;
+
+    int step = 0;
+    int sum = 0;
+
+    LOG.info("Update progress {} changed, {} idle, and {} total to be changed.",
+        scheduled,
+        idle,
+        instanceModificationCount.get());
+
+    if (rollingForward) {
+      while (sum < scheduled && step < maxActiveGroups.size()) {
+        sum += maxActiveGroups.get(step);
+
+        ++step;
+      }
+    } else {
+
+      // Starting with the first step as we're now comparing to idle instead of scheduled
+      // to work backwards from the last step that was executed in the batch steps.
+      sum = maxActiveGroups.get(step);
+
+      // TODO(rdelvalle): Consider if it's necessary to handle fractional steps.
+      // (i.e. halfway between step X and X+1)
```

Comment:
I think now would be the best time to reason about this. I don't think fractional steps are necessary, but maybe you can elaborate what you had in mind.

Reply:
This thought may have come up as a side effect of the old way of calculating the current batch size. What I was thinking about is if half way through a batch size, the scheduler fails over and the update has to continue from where it left off when the next leader is chosen.

E.g: We have batch sizes 10 and 30. Half way through 30 (25 instances updated), the scheduler fails over.

With the new way of calculating which batch size we're on on the fly, I don't think it's an issue anymore.